articles/actions.md

```diff
@@ -0,0 +1,363 @@
+The state machine is only started if the action server *accepts* the goal.
+- **cancel_goal** - Request that the action server stop processing the goal.
+A transition only occurs if the action server *accepts* the request to cancel the goal.
+
+## API
+
+Proposed examples can be in the respository [examples (branch: actions_proposal)](https://github.com/ros2/examples/tree/actions_proposal).
+
+C++:
+
+- [examples/rclcpp/minimal_action_server](https://github.com/ros2/examples/tree/actions_proposal/rclcpp/minimal_action_server)
+- [examples/rclcpp/minimal_action_client](https://github.com/ros2/examples/tree/actions_proposal/rclcpp/minimal_action_client)
+
+Python:
+
+- [examples/rclpy/actions](https://github.com/ros2/examples/tree/actions_proposal/rclpy/actions)
+
+## Middleware implementation
+
+Under the hood, an action is made up of three services and two topics.
+In this section, they are descibed in detail.
+
+![Action Client/Server Interaction Overview](../img/actions/interaction_overview.png)
+
+### Send Goal Service
+
+- **Direction**: client calls server
+- **Request**: description of goal and a UUID for the goal ID
+- **Response**: whether goal was accepted or rejected and the time when the goal was accepted
+
+The purpose of this service is to send a goal to the action server.
+It is the first service called to begin an action, and is expected to return quickly.
+The description of the goal in the request is user-define as part of the [Action Interface Definition](#action-interface-definition).
+
+The QoS settings of this service should be set so the client is guaranteed to receive a response or an action could be executed without a client being aware of it.
+
+### Cancel Goal Service
+
+- **Direction**: client calls server
+- **Request**: goal ID and timestamp
+- **Response**: list of goals that have transitioned to the CANCELING state
+
+The purpose of this service is to request the cancellation of one or more goals on the action server.
+The result indicates which goals will be attempted to be canceled.
+Whether or not a goal is actually canceled is indicated by the status topic and the result service.
+
+The cancel request policy is the same as in ROS 1.
+
+- If the goal ID is empty and timestamp is zero, cancel all goals
+- If the goal ID is empty and timestamp is not zero, cancel all goals accepted at or before the timestamp
+- If the goal ID is not empty and timestamp is zero, cancel the goal with the given ID regardless of the time it was accepted
+- If the goal ID is not empty and timestamp is not zero, cancel the goal with the given ID and all goals accepted at or before the timestamp
+
+### Get Result Service
+
+* **Direction**: client calls server
+* **Request**: goal ID
+* **Response**: status of goal and user defined result
+
+The purpose of this service is to get the final result of a goal.
+After a goal has been accepted the client should call this service to receive the result.
```

Comment:
So can I assume that like other services, there will be a callback to the client when the result is ready?

Reply:
Yes. The idea is to follow a similar pattern as services. Although it is not shown in the proposed minimal example,

https://github.com/ros2/examples/blob/01f24c9c64003be894584c53085c3910ec2268b7/rclcpp/minimal_action_client/not_composable.cpp#L36-L38

I believe we should be able to also provide a callback (like the service API) for responses to goal requests, feedback, and results:

```c++
auto goal_handle = action_client->async_send_goal(goal_msg, goal_response_cb, feedback_cb, result_cb);
```

Reply:
I don't think this needs an additional topic. The client side can make an async result request (e.g. after getting a goal response) and during spin the client can check if a result response has been received, making any user-defined callbacks as needed.